opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -2426,11 +2436,19 @@ namespace Opm
     MultisegmentWell<TypeTag>::
     updateUpwindingSegments()
     {
-        // not considering upwinding for the injectors for now
-        // but we should
-        // and upwinding segment for top segment is itself
         for (int seg = 0; seg < numberOfSegments(); ++seg) {
-            if ( (well_type_ == INJECTOR) || (seg == 0) ) {
+            // special treatment is needed for segment 0
+            if (seg == 0) {
+                if ( (well_type_ == INJECTOR)) {
+                    // maybe it is safer to return -1 to make sure that we don't return the wrong segment number
+                    // for injectors
+                    upwinding_segments_[seg] = seg;
+                    assert(primary_variables_evaluation_[seg][GTotal] >= 0.);
+                    continue;
+                }
+
+                // we don't handle injecting producers.
+                assert(primary_variables_evaluation_[seg][GTotal] <= 0.);
                 upwinding_segments_[seg] = seg;
```

Comment:
This assert is triggered in the 3D_MSW test case. How is injecting producers handled in the rest of the code?

Reply:
For StandardWell it is easier to understand. It is fine to let a producer inject unless THP is involved. For MSW, I do not have an answer yet.  MSW typically needs to handle more complicated fluid behavoir within wellbore.

Reply:
Short answer is that we have not cared too much about whether a producer is injecting for how. Other constraint might recover the well to return to producing.  Not sure whether this is the final design we want.